File: backend/src/main/java/com/optiroute/backend/entity/Vehicle/Tractor.java
package com.optiroute.backend.entity.vehicle;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDate;

import com.optiroute.backend.entity.EntityUtils;
import com.optiroute.backend.type.FuelType;

@Entity
@Table(name = "tractor")
@NoArgsConstructor
@Getter
@Setter
public class Tractor extends EntityUtils {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "external_id")
    private String externalId;

    @Column(name = "external_source")
    private String externalSource = "MANUAL";

    @Column(nullable = false, unique = true, length = 15)
    private String registration;

    @Column(length = 100)
    private String brand;

    @Column(length = 100)
    private String model;

    private Integer maxSpeed;

    @Enumerated(EnumType.STRING)
    @Column(length = 30)
    private FuelType fuelType;

    @Column(precision = 6, scale = 2)
    private BigDecimal averageConsumption;

    @Column(precision = 6, scale = 2)
    private BigDecimal averageConsumptionEmpty;

    private Integer emptyWeightKg;

    private Integer grossCombinationWeightKg;

    private Integer heightCm;

    private Integer widthCm;

    private Integer lengthCm;

    private Integer axleCount;

    @Column(precision = 12, scale = 2)
    private BigDecimal purchaseCost;

    private LocalDate depreciationStartDate;

    private LocalDate depreciationEndDate;

    @Column(name = "is_active", nullable = false)
    private Boolean active = true;
}

File: frontend/src/models/vehicle/Tractor.ts
import type { FuelType } from "@/types/vehicleType"

export interface TractorSummary {
    id: number

    registration: string

    brand: string | null
    model: string | null
}

export interface TractorDetails extends TractorSummary {
    maxSpeed: number | null

    fuelType: FuelType | null
    averageConsumption: number | null
    averageConsumptionEmpty: number | null

    emptyWeightKg: number | null
    grossCombinationWeightKg: number | null

    heightCm: number | null
    widthCm: number | null
    lengthCm: number | null

    axleCount: number | null

    purchaseCost: number | null

    depreciationStartDate: string | null
    depreciationEndDate: string | null

    externalId: string | null
    externalSource: string

    createdAt: string
    updatedAt: string
}

export interface CreateTractorRequest {
    externalId: string | null
    externalSource: string

    registration: string

    brand: string | null
    model: string | null

    maxSpeed: number | null

    fuelType: FuelType | null
    averageConsumption: number | null
    averageConsumptionEmpty: number | null

    emptyWeightKg: number | null
    grossCombinationWeightKg: number | null

    heightCm: number | null
    widthCm: number | null
    lengthCm: number | null

    axleCount: number | null

    purchaseCost: number | null

    depreciationStartDate: string | null
    depreciationEndDate: string | null
}

export interface UpdateTractorRequest {
    registration: string

    brand: string | null
    model: string | null

    maxSpeed: number | null

    fuelType: FuelType | null
    averageConsumption: number | null
    averageConsumptionEmpty: number | null

    emptyWeightKg: number | null
    grossCombinationWeightKg: number | null

    heightCm: number | null
    widthCm: number | null
    lengthCm: number | null

    axleCount: number | null

    purchaseCost: number | null

    depreciationStartDate: string | null
    depreciationEndDate: string | null
}

export interface TractorFormData {
    registration: string
    brand: string | null
    model: string | null

    maxSpeed: number | null

    fuelType: FuelType | null
    averageConsumption: number | null
    averageConsumptionEmpty: number | null

    emptyWeightKg: number | null
    grossCombinationWeightKg: number | null

    heightCm: number | null
    widthCm: number | null
    lengthCm: number | null

    axleCount: number | null

    purchaseCost: number | null
    depreciationStartDate: string | null
    depreciationEndDate: string | null
}